File: lib/src/core/change_detection/proto_record.dart
library angular2.src.core.change_detection.proto_record;

import "binding_record.dart" show BindingRecord;
import "directive_record.dart" show DirectiveIndex;

enum RecordType {
  Self,
  Const,
  PrimitiveOp,
  PropertyRead,
  PropertyWrite,
  Local,
  InvokeMethod,
  InvokeClosure,
  KeyedRead,
  KeyedWrite,
  Pipe,
  Interpolate,
  SafeProperty,
  CollectionLiteral,
  SafeMethodInvoke,
  DirectiveLifecycle,
  Chain,
  SkipRecordsIf,
  SkipRecordsIfNot,
  SkipRecords
}

class ProtoRecord {
  RecordType mode;
  String name;
  var funcOrValue;
  List<dynamic> args;
  List<dynamic> fixedArgs;
  num contextIndex;
  DirectiveIndex directiveIndex;
  num selfIndex;
  BindingRecord bindingRecord;
  bool lastInBinding;
  bool lastInDirective;
  bool argumentToPureFunction;
  bool referencedBySelf;
  num propertyBindingIndex;
  ProtoRecord(
      this.mode,
      this.name,
      this.funcOrValue,
      this.args,
      this.fixedArgs,
      this.contextIndex,
      this.directiveIndex,
      this.selfIndex,
      this.bindingRecord,
      this.lastInBinding,
      this.lastInDirective,
      this.argumentToPureFunction,
      this.referencedBySelf,
      this.propertyBindingIndex) {}
  bool isPureFunction() {
    return identical(this.mode, RecordType.Interpolate) ||
        identical(this.mode, RecordType.CollectionLiteral);
  }

  bool isUsedByOtherRecord() {
    return !this.lastInBinding || this.referencedBySelf;
  }

  bool shouldBeChecked() {
    return this.argumentToPureFunction ||
        this.lastInBinding ||
        this.isPureFunction() ||
        this.isPipeRecord();
  }

  bool isPipeRecord() {
    return identical(this.mode, RecordType.Pipe);
  }

  bool isConditionalSkipRecord() {
    return identical(this.mode, RecordType.SkipRecordsIfNot) ||
        identical(this.mode, RecordType.SkipRecordsIf);
  }

  bool isUnconditionalSkipRecord() {
    return identical(this.mode, RecordType.SkipRecords);
  }

  bool isSkipRecord() {
    return this.isConditionalSkipRecord() || this.isUnconditionalSkipRecord();
  }

  bool isLifeCycleRecord() {
    return identical(this.mode, RecordType.DirectiveLifecycle);
  }
}

File: es6/prod/src/core/change_detection/proto_record.js
export var RecordType;
(function (RecordType) {
    RecordType[RecordType["Self"] = 0] = "Self";
    RecordType[RecordType["Const"] = 1] = "Const";
    RecordType[RecordType["PrimitiveOp"] = 2] = "PrimitiveOp";
    RecordType[RecordType["PropertyRead"] = 3] = "PropertyRead";
    RecordType[RecordType["PropertyWrite"] = 4] = "PropertyWrite";
    RecordType[RecordType["Local"] = 5] = "Local";
    RecordType[RecordType["InvokeMethod"] = 6] = "InvokeMethod";
    RecordType[RecordType["InvokeClosure"] = 7] = "InvokeClosure";
    RecordType[RecordType["KeyedRead"] = 8] = "KeyedRead";
    RecordType[RecordType["KeyedWrite"] = 9] = "KeyedWrite";
    RecordType[RecordType["Pipe"] = 10] = "Pipe";
    RecordType[RecordType["Interpolate"] = 11] = "Interpolate";
    RecordType[RecordType["SafeProperty"] = 12] = "SafeProperty";
    RecordType[RecordType["CollectionLiteral"] = 13] = "CollectionLiteral";
    RecordType[RecordType["SafeMethodInvoke"] = 14] = "SafeMethodInvoke";
    RecordType[RecordType["DirectiveLifecycle"] = 15] = "DirectiveLifecycle";
    RecordType[RecordType["Chain"] = 16] = "Chain";
    RecordType[RecordType["SkipRecordsIf"] = 17] = "SkipRecordsIf";
    RecordType[RecordType["SkipRecordsIfNot"] = 18] = "SkipRecordsIfNot";
    RecordType[RecordType["SkipRecords"] = 19] = "SkipRecords"; // Skip records unconditionally
})(RecordType || (RecordType = {}));
export class ProtoRecord {
    constructor(mode, name, funcOrValue, args, fixedArgs, contextIndex, directiveIndex, selfIndex, bindingRecord, lastInBinding, lastInDirective, argumentToPureFunction, referencedBySelf, propertyBindingIndex) {
        this.mode = mode;
        this.name = name;
        this.funcOrValue = funcOrValue;
        this.args = args;
        this.fixedArgs = fixedArgs;
        this.contextIndex = contextIndex;
        this.directiveIndex = directiveIndex;
        this.selfIndex = selfIndex;
        this.bindingRecord = bindingRecord;
        this.lastInBinding = lastInBinding;
        this.lastInDirective = lastInDirective;
        this.argumentToPureFunction = argumentToPureFunction;
        this.referencedBySelf = referencedBySelf;
        this.propertyBindingIndex = propertyBindingIndex;
    }
    isPureFunction() {
        return this.mode === RecordType.Interpolate || this.mode === RecordType.CollectionLiteral;
    }
    isUsedByOtherRecord() { return !this.lastInBinding || this.referencedBySelf; }
    shouldBeChecked() {
        return this.argumentToPureFunction || this.lastInBinding || this.isPureFunction() ||
            this.isPipeRecord();
    }
    isPipeRecord() { return this.mode === RecordType.Pipe; }
    isConditionalSkipRecord() {
        return this.mode === RecordType.SkipRecordsIfNot || this.mode === RecordType.SkipRecordsIf;
    }
    isUnconditionalSkipRecord() { return this.mode === RecordType.SkipRecords; }
    isSkipRecord() {
        return this.isConditionalSkipRecord() || this.isUnconditionalSkipRecord();
    }
    isLifeCycleRecord() { return this.mode === RecordType.DirectiveLifecycle; }
}
